src/main/java/net/minecraftforge/registries/ForgeRegistry.java

```diff
@@ -101,6 +106,42 @@
     private final ResourceLocation name;
     private final RegistryKey<Registry<V>> key;
     private final RegistryBuilder<V> builder;
+    
+    private final Codec<V> CODEC = new Codec<V>() {
+        @Override
+        public <T> DataResult<Pair<V, T>> decode(DynamicOps<T> ops, T input) {
+            if(ops.compressMaps())
+            {
+                return ops.getNumberValue(input).flatMap(n ->
+                {
+                    int id = n.intValue();
+                    if(ids.get(id) == null)
+                    {
+                        return DataResult.error("Unknown registry id " + n);
+                    }
+                    V val = getValue(id);
+                    return DataResult.success(val);
+                }).map(v -> Pair.of(v, input));
+            }
+            else
+            {
+                return ResourceLocation.CODEC.decode(ops, input).flatMap(keyValuePair -> !ForgeRegistry.this.containsKey(keyValuePair.getFirst()) ?
+                        DataResult.error("Unknown registry key: " + keyValuePair.getFirst()) :
+                        DataResult.success(keyValuePair.mapFirst(ForgeRegistry.this::getValue)));
+            }
+        }
+
+        @Override
+        public <T> DataResult<T> encode(V input, DynamicOps<T> ops, T prefix) {
+            ResourceLocation key = getKey(input);
+            if(key == null)
+            {
+                return DataResult.error("Unknown registry element " + input);
+            }
+            T toMerge = ops.compressMaps() ? ops.createInt(getID(input)) : ops.createString(key.toString());
```

Comment:
Question: why is the `ResourceLocation` converted to a string via `.toString()` instead of its `Codec` (from `ResourceLocation.CODEC`)? (aside from this being a formatted copy from vanilla's `Registry#encode`)
(Yes, I understand this works as vanilla uses it, but I'm curious as to why not the codec too; disclaimer: I have only a passing familiarity with codecs)

Reply:
As you sait it's a copy of the vanilla registry implementation. But just look at the ResourceLocation.CODEC and you will see that it's encode method is basically ResourceLocation::toString.
So there is no need of the ResoureceLocation codec in this case.
Also the codec would return a DataResult<T> with T being the String so I would need to add severals checks to handle the case the codec failed to encode the ResourceLocation, this should not happen because the encode is just the toString() method but still the code for the checks would be needed to retrieve the encoded String from the DataResult.